fix for #38585: Signed primary columns changed value

KrasnoshchokBohdan opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
changed "unsigned" attribute for "value_id" column for tables from original list (This is applicable to the following tables:... from author post) + catalog_category_entity_datetime, customer_address_entity_datetime
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#38585

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. mysql -u root -p
2. use (your magento2 db name)
3. desc catalog_product_entity_text (or any table from the list)
4. check "value_id" type, should be unsigned
<img width="567" alt="image" src="https://github.com/user-attachments/assets/223eb2cc-40b8-464b-9c15-d7d8e666d4dc">


### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

DanK894 committed:
magento/magento2#38585: Database tables with primary column having signed instead of unsigned definition which is loss of usable values

- changed unsigned attribute for value_id column for tables from original list + catalog_category_entity_datetime customer_address_entity_datetime

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan commented:
@magento run Functional Tests B2B Functional Tests CE Functional Tests EE

KrasnoshchokBohdan commented:
@magento run Functional Tests B2B

KrasnoshchokBohdan commented:
@magento run Functional Tests B2B

KrasnoshchokBohdan commented:
@magento run Functional Tests CE

KrasnoshchokBohdan commented:
@magento run Functional Tests EE

KrasnoshchokBohdan commented:
Failed tests seems flaky to me

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-38585

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented:
We are moving this PR `On Hold` as some internal discussion is going on related to it.

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-38585

engcom-Hotel commented:
@magento run all tests

engcom-Bravo committed:
Merge branch '2.4-develop' into fix-for-issue-38585

engcom-Bravo commented:
@magento run all tests

engcom-Bravo commented:
Hi @KrasnoshchokBohdan,

Thanks for the collaboration & contribution!

### :heavy_check_mark: QA Passed


**Preconditions:**

- Install fresh Magento **`2.4-develop`**

**Steps to reproduce**

- mysql -u root -p
- use (your magento2 db name)
- desc catalog_product_entity_text (or any table from the list)
- check "value_id" type, should be unsigned


**Before:** :heavy_multiplication_x: 


<img width="1402" alt="Screenshot 2024-09-11 at 11 30 08" src="https://github.com/user-attachments/assets/bd2b62ef-ba26-459c-bd90-d045acf11837">




**After:** :heavy_check_mark: 

<img width="1402" alt="Screenshot 2024-09-11 at 14 34 59" src="https://github.com/user-attachments/assets/7457da4d-179e-4c19-92f7-fb3729e33673">



Builds are failed. Hence, moving this PR to **Extended Testing**.

Thanks

engcom-Charlie committed:
Merge branch '2.4-develop' into fix-for-issue-38585

engcom-Charlie commented:
@magento run all tests

engcom-Charlie commented:
The functional B2B failures are not same in 2 recent builds on same code. The failing tests are not failing because of PR changes, seems to be flaky. Hence moving this PR to Merge in Progress.

**Run 1:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38920/fb2f4094d46ade96a80c9e3c3b13d81e/Functional/allure-report-b2b/index.html

<img width="1239" alt="image" src="https://github.com/user-attachments/assets/62131eb1-4266-4f50-8c79-03154968cb66">


**Run 2:** 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38920/aa6286563162120e7ba890719aa5fb99/Functional/allure-report-b2b/index.html#categories/9d8166a0c03ff30be1e72531365d31ec/45b8b3964c62d0b0/
<img width="1131" alt="image" src="https://github.com/user-attachments/assets/9e40bcd8-a538-4bd6-8982-1e466b31bfdc">

ihor-sviziev commented:
Changes look good to me, but just wondering if we should update all such values at once?
https://github.com/search?q=repo%3Amagento%2Fmagento2+unsigned%3D%22false%22+path%3A*.xml&type=code

KrasnoshchokBohdan commented:
@ihor-sviziev I can't think of any reasons why it would be better to split it up

ihor-sviziev reviewed (changes requested):
Could you please review all similar places in other modules and fix them all together?

KrasnoshchokBohdan commented:
@ihor-sviziev but this is all i found
https://github.com/search?q=repo%3Amagento%2Fmagento2+unsigned%3D%22false%22+path%3A*.xml&type=code

or, maybe, I misunderstand you, please add additional information

ihor-sviziev commented:
@KrasnoshchokBohdan that's the list that I found:
- https://github.com/magento/magento2/blob/ea3216321185052523ad65728ee97a53fb8bdb86/app/code/Magento/Cms/etc/db_schema.xml#L11
- https://github.com/magento/magento2/blob/ea3216321185052523ad65728ee97a53fb8bdb86/app/code/Magento/Tax/etc/db_schema.xml#L21
- https://github.com/magento/magento2/blob/ea3216321185052523ad65728ee97a53fb8bdb86/app/code/Magento/LoginAsCustomerLog/etc/db_schema.xml#L10
Could you please check?

KrasnoshchokBohdan committed:
Merge branch 'magento:2.4-develop' into fix-for-issue-38585

DanK894 committed:
magento/magento2#38585: Database tables with primary column having signed instead of unsigned definition which is loss of usable values

- changed unsigned attribute for primary columns

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan commented:
@ihor-sviziev thank you! I hope I understood your point

if we check the original issue description, we will see that author has pointed to the primary columns with name "value_id", so I focused on them, my bad.

I would also like to note that the search in php-storm does not work 100% correctly, I tried to check carefully, but I am afraid that I might have missed some files, so please let me know if there are any - I fix it

ihor-sviziev commented:
@KrasnoshchokBohdan, could you check test failures?
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38920/a958c94246b4436e20c0578b455ffce8/Integration/console-error-logs.html
It might be not a good idea to fix it inall places

engcom-Charlie commented:
Hi @KrasnoshchokBohdan,

Thank you for your contribution!

The PR have been moved to Merge in Progress after approval, testing and build stabilization but @KrasnoshchokBohdan, as you are started modifying the code base, moving this PR back to Changes Requested again. 

Please let us know once you are completed the development on it.

Thank you!

DanK894 committed:
magento/magento2#38585: Database tables with primary column having signed instead of unsigned definition which is loss of usable values

- rollback of unsigned attribute for main columns in app/code/Magento/Tax/etc/db_schema.xml due to test errors

KrasnoshchokBohdan commented:
@magento run all tests

DanK894 committed:
magento/magento2#38585: Database tables with primary column having signed instead of unsigned definition which is loss of usable values

- rollback of unsigned attribute for main columns in app/code/Magento/Cms/etc/db_schema.xml due to test errors

KrasnoshchokBohdan commented:
@magento run Integration Tests

KrasnoshchokBohdan commented:
@magento run Database Compare

KrasnoshchokBohdan commented:
@magento run Unit Tests Static Tests WebAPI Tests

KrasnoshchokBohdan commented:
@magento run Unit Tests

KrasnoshchokBohdan commented:
@magento run WebAPI Tests

KrasnoshchokBohdan commented:
@magento run Magento Health Index

KrasnoshchokBohdan commented:
@magento run Static Tests Semantic Version Checker

KrasnoshchokBohdan commented:
@magento run Static Tests, Semantic Version Checker

KrasnoshchokBohdan commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

KrasnoshchokBohdan commented:
@ihor-sviziev @engcom-Charlie 

Hi! Now it seems to look better, what do you think?

ihor-sviziev reviewed (changes requested):
@KrasnoshchokBohdan, I just reviewed this PR again, and see that by adjusting type of the columns we were getting issues like this:
```
  Column definition "product_tax_class_id" and reference column definition "c
  lass_id" are different in tables "tax_calculation" and "tax_class"
  Column definition "customer_tax_class_id" and reference column definition "
  class_id" are different in tables "tax_calculation" and "tax_class"
  Column definition "tax_calculation_rate_id" and reference column definition
   "tax_calculation_rate_id" are different in tables "tax_calculation" and "t
  ax_calculation_rate"
  Column definition "tax_calculation_rule_id" and reference column definition
   "tax_calculation_rule_id" are different in tables "tax_calculation" and "t
  ax_calculation_rule"
  Column definition "tax_calculation_rate_id" and reference column definition
   "tax_calculation_rate_id" are different in tables "tax_calculation_rate_ti
  tle" and "tax_calculation_rate"


```
You've reverted these changes, which is great, thanks!

I am wondering if people might have similar issues with other tables that you've changed just because they have foreign keys to these tables in custom extensions, and Magento doesn't have those custom extensions.

I believe the number of potentially affected people will be higher than the number of people getting to the issue you're trying to resolve. Those who need it can apply these changes as a patch and verify that they do not affect their Magento installation. 

Having that in mind, I don't know if we should accept this pull request.
Please give us your thoughts on this.

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-38585

engcom-Hotel commented:
@magento run all tests

rostilos commented:
@ihor-sviziev 
Perhaps to maintain compatibility there is an option to create a datapatch instead of directly modifying the schema DB?

ihor-sviziev commented:
@rostilos, unfortunately it won’t help because of dependencies in 3rt party extensions. These extensions still will have a db schema files, which declare the current state. I don’t think it’s possible to make backward compatible version at all, while benefit from this change will be for pretty small amount for installations

engcom-Charlie commented:
Hi @engcom-Hotel,

Can you please go through @ihor-sviziev's [comment](https://github.com/magento/magento2/pull/38920#pullrequestreview-2316448365) and review this PR again to decide whether we can proceed ahead with this PR or not.

Thank you!

engcom-Hotel commented:
Hello @KrasnoshchokBohdan,

I agree with @ihor-sviziev's input here. This PR may impact some third-party extension that uses these tables as foreign keys.

I think we can close this PR.

Please suggest @KrasnoshchokBohdan @rostilos @ihor-sviziev.

Thanks

engcom-Hotel commented:
We are closing this PR as the fix may impact the 3rd party extensions.

hostep commented:
I heard some whispers a few months ago about a potential upcoming Magento 2.5.0 version, so maybe if that actually happens, we can open this PR again, because it's a new major version where backwards compatibility breaking is allowed

rostilos commented:
@engcom-Hotel 
Perhaps, given the post above, we can reopen it and assign an appropriate label?

ihor-sviziev commented:
I don't mind re-opening it for 2.5, but I didn't hear any news about it

hostep commented:
We have more and more shops that started like 10 years ago on Magento 2 that slowly start reaching this signed 32 bit limit on some of these tables (mostly happens on `catalog_product_entity_int` for us it seems). We already have 2 shops in the past few months that reached it and that we need to patch now to keep things running smoothly for them.

I'm wondering how backwards incompatible this change would be like is being mentioned above a few times, I understand that if a module needs to support multiple Magento versions and wants to reference one of these columns, that it's hard to support both a `signed` and `unsigned` reference depending on Magento version, but wouldn't then the solution be to use `unsigned` all the time, even for older Magento versions that the module supports? Or will `bin/magento setup:upgrade` (or mysql) start yelling when that happens?
It's probably worth giving this a try to see how this would behave in real life instead of theorising (I'll see if I can find some time for this in the near future), because if we can tell 3rd party modules to always use `unsigned` on their end when referencing some of these columns, I think we can move forward with this change here and it wouldn't cause a big headache for most vendors.
